Assets

rowanseymour opened the pull request:
An app for restricted access to files in storage. Provides 3 different types of URL for assets:
 * Landing pages: http://localhost:8000/assets/download/contact_export/1234/
 * Direct downloads: http://localhost:8000/assets/stream/contact_export/1234/
 * API access: http://localhost:8000/api/v1/assets?type=contact_export&identifier=1234

norkans7 commented on temba/contacts/models.py:
```diff
@@ -1468,20 +1464,17 @@ def add_sheet(book, sheet_number, fields):
+        from temba.assets import AssetType
+        handler = AssetType.contact_export.handler
+        handler.save(self.pk, File(temp), 'csv' if use_csv else 'xls')
 
```
we can also use the assets url for link sent in the email

rowanseymour replied:
The URLs in emails currently point to _/org/download/contacts/xxx_ which I think we'd want to keep as it provides a landing page.

norkans7 replied:
we could adjust AssetView to extend SmartTemplateView and add the template for it

also we can translate the old links to the new ones by redirecting their view

norkans7 commented on temba/assets/handlers.py:
```diff
@@ -0,0 +1,156 @@
+    extensions = ('xls', 'csv')
+
+
+def has_org_permission(org, user, permission):
```
we have a similar in OrgPermsMixin

ericnewcomer commented:
Another thing to consider is that the recordings that are part of a flow definition (the prompts by the flow, not the contact recordings). When somebody exports this definition and shares it we won't be able to pull in those prompts on a different org/site if they are gated by this.

rowanseymour commented:
* Adding landing page for asset downloads
* Existing org/downloads/xx URLs now redirect to equivalent asset URLs
* Converted all three export types: contact, message and results

Have not added a migration to remove the filename field from the export task models. Would prefer to do that after verifying that all the files were correctly re-organised.

rowanseymour commented on temba/settings_common.py:
```diff
@@ -33,6 +33,7 @@
 
 # where recordings and exports are stored
 AWS_STORAGE_BUCKET_NAME = 'dl.temba.io'
+STORAGE_ROOT_DIR = 'test_orgs' if TESTING else 'orgs'
```
Let's us keep files generated by unit tests separate from others

ericnewcomer commented on temba/assets/handlers.py:
```diff
@@ -0,0 +1,151 @@
+    Asset file could not be found
+    """
+    pass
+
```
Don't really need all these new Exception classes since they aren't explicitly caught/handled anywhere. Why not just just 
```
raise Exception(_('Asset file could not be found')) 
```
...like we do everywhere else? Better to keep our approach here consistent.

rowanseymour replied:
We translate them into different error messages for the landing page and for direct downloads. I also don't think we should be raising generic `Exception` instances - should at least be using subclasses like `ValueError`.

rowanseymour replied:
Actually they also get converted to HTTP status codes in the latter case, so need some way of differentiating between error types.

ericnewcomer commented on temba/assets/handlers.py:
```diff
@@ -0,0 +1,151 @@
+                return path
+
+        raise AssetFileNotFound()
+
```
Nice, really like how DRY this is! Can we ditch handlers.py and move these methods into models? The notion of handlers is already pretty overloaded and don't to introduce another file just for this.

nicpottier replied:
Confession, I first recoiled at the idea of handlers.py and the general architecture here, but I've come around. I agree with Eric that the main confusion is probably around naming. Handler sounds a lot like something that is dealing with requests and these are doing a bit less than that really, they are more about storing and retrieving things.

How about renaming handlers.py to models.py, just because that's where we put business logic kind of stuff, and then using Store instead of Handler? I think that makes it super obvious what these do without introducing the ambiguity as to whether they actually handle requests or the like.

But ya, great job on the architecture of this, looks good.

rowanseymour replied:
Renamed to `Store` and moved into _models.py_. `Handler` was pretty lazy on the naming front.

norkans7 commented on temba/api/views.py:
```diff
@@ -2585,6 +2588,27 @@ def get_read_explorer(cls):
+        if not type_name or not identifier:
+            return HttpResponseBadRequest("Must provide type and identifier")
+
+        if type_name not in AssetType.__members__:
```
what about recordings here?

rowanseymour replied:
Is your concern that a single identifier won't be enough to reference a recording? Maybe we use a composite identifier...

We're not implementing this for recordings just yet, and this endpoint isn't documented so we can change it later.

norkans7 replied:
my concern was that I do not see it implemented

nicpottier commented:
This good to go now Rowan?

rowanseymour commented:
I've only been able to test the migrations with a few files locally so I was thinking about setting the staging server up to use S3 storage and testing them there with a bit more data. Or we could just push it...

nicpottier commented:
Ya, would definitely like to see it working on staging first. Might be nice to have staging working against it's own S3 backend anyways instead of having it walk over prod.

ericnewcomer commented:
I'm not necessarily opposed the exception model here, just advocating consistency. We should powwow on our convention there.

rowanseymour commented:
Love a good powpow! So I think there are cases like this where multiple custom exception types are useful for conveying different error states that need to be handled differently. We definitely don't need to go creating a custom exception type for everything, and most of the time something like `ValueError` is appropriate. 

Regarding convention and consistency - I think we should try to avoid raising generic `Exception` instances as that forces calling code to catch `Exception` - which at least with PyCharm's defaults, earns you a warning for catching too broad an exception. We should definitely avoid blanket excepts (`except:`) as these catch all sorts like `KeyboardInterrupt` etc.

nicpottier commented:
POWOW!!!111

I think you are right Rowan that we probably shouldn't be using Exception as much, I hadn't thought of it but you are absolutely correct that we probably are trapping kill signals that we shouldn't be and that could be causing restart problems now and then.

I don't really want to have a zillion exception types either though, so maybe we should add a TembaException() and use localized strings with it in all cases except where a different exception type is needed because of a different code path on the handling end.

I was originally thinking we could probably reuse existing exception classes in a lot of cases. (there must be built in Python IO exceptions that capture access defined and not found for example?) But on reflection it seems that better to have all OUR exceptions to have a common ancestor. That said, maybe some of these common ones like "NotFound" or "AccessDenied" should be more generic and belong in .utils.

But in this case, what about reducing this just to two exceptions types (it does seem Python standard is having it end in Exception or Error): AssetNotFoundException(TembaException) and AssetAccessDeniedException(TembaException) each with localized strings when they are raised.

So a proposed "standard" for exception handling might be:
 1. Don't raise Exception, raise TembaException instead (in .utils?)
 2. Only create new TembaException subclasses when needed for different code paths by caller, by default, use localized strings to further describe exception (and so that something useful can be bubbled to user)
 3. Exception type names should end in Exception

Thoughts?

rowanseymour commented:
Perhaps worth listing out how different types of exceptions are used - I can think of at least 3 different ways that we use them:

1. As sanity checks for us - often these are `ValueError` instances with no localization, and `# pragma: no cover`. They don't ever become user facing but they might end up in our logs.
2. As something to be translated into a user facing error message - for example `ValidationError` instances in forms and API. Users see them, so they need localized messages.
3. For internal flow - we raise and catch these internally, thus there's not much use for them to have error messages. For example the asset related ones in this PR and `EvaluationError` in the expression parser.

If we introduced our own exception base class then we could use it for case #1 (though not much to be gained since `ValueError` does the job just fine). For case #3 we could use it as a base for exceptions like those in the assets app. Again though.. am not seeing much benefit.

Regarding naming, PEP8 says exceptions should _Error_ name suffix (https://www.python.org/dev/peps/pep-0008/#exception-names) and it is 4 less characters to type...

nicpottier commented:
Error it is then, all for shorter.

I think using TembaError as a base (and generic error) just makes it easier to do catch alls that don't risk catching system errors. I'm pretty sure you could simplify (IE, have less lines of code and logic) your case if you just used two classes with localized error messages.

There is a place where we shouldn't be using TembaError, and that's with form validation, since those are actually caught and parsed by Django/Smartmin. There we should stick to ValidationError or whatever it is..

ericnewcomer commented:
Ya, I'm for using TembaError in general just so that we know when we are catching ourselves and not a package we use. Having a handful of standard Errors we use living in utils sounds spiffy. Perhaps with have localized default messages.

Also, by powwow, I was thinking we'd chat on our call and agree on a convention. But I guess this works!

rowanseymour commented:
Back to the specifics of this PR for a moment - I've updated the migrations so that if the file copy is successful, then task.filename as nulled. That gives us a definitive way to list any files that didn't copy for some unforeseen reason. Next step (not included here) would be a simple migration to remove the filename field on all export task models.

I changed TextIt Staging to use S3, exported a bunch of things and tested it there. All worked fine, though obviously it didn't copy files saved to media/ before the server was switched to S3.

nicpottier commented:
Let's do this thing!
